api/envoy/config/rbac/v2alpha/rbac.proto

```diff
@@ -0,0 +1,172 @@
+ repeated Permission permissions = 1 [(validate.rules).repeated .min_items = 1];
+
+ // Required. List of principals that are assigned/denied the role based on “action_type”.
+ repeated Principal principals = 2 [(validate.rules).repeated .min_items = 1];
+}
+
+message StringMatch {
+  oneof match_pattern {
+    // The input string must match exactly the string specified here.
+    // Or it is a "*", which means that it matches any string.
+    string simple = 1;
+
+    // The input string must have the prefix specified here.
+    string prefix = 2;
+
+    // The input string must have the suffix specified here.
+    string suffix = 3;
+
+    // The input string must match the regular expression specified here.
+    // The regex grammar is defined `here
+    // <http://en.cppreference.com/w/cpp/regex/ecmascript>`_.
+    string regex = 4;
+ }
+}
+
+// Specifies how to match an entry in a map.
+message MapEntryMatch {
+  // The key to select an entry from the map.
+  string key = 1;
+
+  // A list of matched values.
+  repeated StringMatch values = 2;
+}
+
+// Specifies how to match IP addresses.
+message IpMatch {
+  // IP addresses in CIDR notation.
+  repeated envoy.api.v2.core.CidrRange cidrs = 1;
+}
+
+// Specifies how to match ports.
+message PortMatch {
+  // Port numbers.
+  repeated uint32 ports = 1;
+}
+
+// Permission defines a permission to access the service.
+message Permission {
+  // Optional. A list of HTTP paths or gRPC methods.
+  // gRPC methods must be presented as fully-qualified name in the form of
+  // packageName.serviceName/methodName.
+  // If this field is unset, it applies to any path.
+  repeated StringMatch paths = 1;
+
+  // Required. A list of HTTP methods (e.g., "GET", "POST").
+  // If this field is unset, it applies to any method.
+  repeated string methods = 2;
+
+  // Definition of a custom condition.
+  message Condition {
+    oneof condition_spec {
+      // Header match. This matches to the "request.http.headers" field in
+      // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
+      // The map key is the header name. The header specifies how the service is accessed.
+      MapEntryMatch header = 1;
+
+      // Destination IP addresses.
+      IpMatch destination_ips = 2;
+
+      // Destination ports.
+      PortMatch destination_ports = 3;
+    }
+  }
+
+  // Optional. Custom conditions.
+  repeated Condition conditions = 3;
+}
+
+// Principal defines an identity or a group of identities.
+message Principal {
+  // Authentication attributes for principal. These could be filled out inside RBAC filter.
+  // Or if an authentication filter is used, they can be provided by the authentication filter.
+  message Authenticated {
+    // Optional. The name of the principal. This matches to the "source.principal" field in
+    // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
+    // If unset, it applies to any user.
+    string name = 1;
```

Comment:
Can we make either the `name` field repeated? Otherwise it's hard to convert from the ServiceRoleBinding `user` field that is included in a repeated [subjects](https://github.com/istio/api/blob/793db0cb1a2e33012166f1d6ee5918596b8199d4/rbac/v1alpha1/rbac.proto#L164) field.

Reply:
Subject is renamed as "Principal" here. Each policy has permissions (role) and principals (subjects). Note that Principal is already repeated in Policy.

Reply:
Ah, I see, you're right, thanks for the explanation.